src/vs/workbench/contrib/terminalContrib/accessibility/browser/terminal.accessibility.contribution.ts

```diff
@@ -48,56 +53,219 @@ class TextAreaSyncContribution extends DisposableStore implements ITerminalContr
 
-class AccessibleBufferContribution extends DisposableStore implements ITerminalContribution {
-	static readonly ID = 'terminal.accessible-buffer';
-	private _xterm: IXtermTerminal & { raw: Terminal } | undefined;
-	static get(instance: ITerminalInstance): AccessibleBufferContribution | null {
-		return instance.getContribution<AccessibleBufferContribution>(AccessibleBufferContribution.ID);
-	}
-	private _accessibleBufferWidget: AccessibleBufferWidget | undefined;
 
+export class TerminalAccessibleBufferProvider extends DisposableStore implements IAccessibleContentProvider {
+	options: IAccessibleViewOptions = { type: AccessibleViewType.View };
+	verbositySettingKey = AccessibilityVerbositySettingId.Terminal;
+	private _xterm: IXtermTerminal & { raw: Terminal } | undefined;
 	constructor(
-		private readonly _instance: ITerminalInstance,
-		processManager: ITerminalProcessManager,
-		widgetManager: TerminalWidgetManager,
-		@IInstantiationService private readonly _instantiationService: IInstantiationService,
-		@IConfigurationService configurationService: IConfigurationService
+		private readonly _instance: Pick<ITerminalInstance, 'onDidRunText' | 'focus' | 'shellType' | 'capabilities' | 'onDidRequestFocus' | 'resource'>,
+		private _bufferTracker: BufferContentTracker,
+		@IModelService _modelService: IModelService,
+		@IConfigurationService _configurationService: IConfigurationService,
+		@IContextKeyService _contextKeyService: IContextKeyService,
+		@ITerminalService _terminalService: ITerminalService,
+		@IConfigurationService configurationService: IConfigurationService,
+		@IAccessibleViewService private readonly _accessibleViewService: IAccessibleViewService
 	) {
 		super();
 		this.add(_instance.onDidRunText(() => {
 			const focusAfterRun = configurationService.getValue(TerminalSettingId.FocusAfterRun);
 			if (focusAfterRun === 'terminal') {
 				_instance.focus(true);
 			} else if (focusAfterRun === 'accessible-buffer') {
-				this.show();
+				_accessibleViewService.show(this);
 			}
 		}));
+		this.registerListeners();
 	}
-	layout(xterm: IXtermTerminal & { raw: Terminal }): void {
-		this._xterm = xterm;
+
+	onClose() {
+		this._instance.focus();
 	}
-	async show(): Promise<void> {
+	registerListeners(): void {
 		if (!this._xterm) {
 			return;
 		}
-		if (!this._accessibleBufferWidget) {
-			this._accessibleBufferWidget = this.add(this._instantiationService.createInstance(AccessibleBufferWidget, this._instance, this._xterm));
+		this._xterm.raw.onWriteParsed(async () => {
+			if (this._xterm!.raw.buffer.active.baseY === 0) {
+				this.provideContent();
+				this._accessibleViewService.show(this);
+			}
+		});
+		const onRequestUpdateEditor = Event.latch(this._xterm.raw.onScroll);
+		this.add(onRequestUpdateEditor(() => this._accessibleViewService.show(this)));
+	}
+
+	provideContent(): string {
+		this._bufferTracker.update();
+		return this._bufferTracker.lines.join('\n');
+	}
```

Comment:
Depending on how often this is called, this may end up consuming and maybe not releasing a lot of memory.

Reply:
Yes good point. I should save the contents, just return them here, and only update them under certain circumstances.

Reply:
How would I fix that? caching the current lines and removing/ adding to them instead of recreating each time?

Reply:
See my new comment